### PR TITLE
Expand Q space search radius in DetectorSearch to counter missing peaks in PredictPeaks

### PR DESCRIPTION
**Pull request by KedoKudo**
**Description of work.**

- Original Gitlab defect: [SCD150](https://code.ornl.gov/sns-hfir-scse/diffraction/single-crystal/single-crystal-diffraction/-/issues/150#note_580553)
- This is #31767 into `master`

The current default q space search radius settings in DetectorSearch will lead to some missing peaks when trying to use PredictPeaks on data collected with WAND2@ORNL.
This PR expand the search radius to cover a larger q space so that the missing peaks can be correctly associated with detector, therefore no longer "missing".

**To test:**
Since the testing data is only available on analysis, the testing needs to be performed on the analysis cluster.
- build the branch and launch the workbench
- use the following code the process the data on the net-drive
```python
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

# data IPTS and run numbers ----------------------------------------------------
data_ipts = 25156
data_runs = [543722, 545522]  #[325670, 327470]

# vandium IPTS and run number --------------------------------------------------
van_ipts = 23858
van_run = 531821  #320259

# wavelength (angstrom) --------------------------------------------------------
wavelength = 1.486

# minimum and maximum values for Q sample --------------------------------------
min_values = [-7.5, -0.65, -4.4]
max_values = [6.8, 0.65, 7.5]

# obliquity parallax coefficient
cop = 1.0

# detector grouping for loading data faster ------------------------------------
grouping = '4x4'  # '2x2', 'None', '4x4'

# ---

# predict peak centering -------------------------------------------------------
reflection_condition = 'Primitive'

# ---

norm = LoadWANDSCD(IPTS=van_ipts, RunNumbers=van_run, Grouping=grouping)

data = LoadWANDSCD(IPTS=data_ipts,
                   RunNumbers='{}-{}'.format(*data_runs),
                   Grouping=grouping)

norm_replicated = ReplicateMD(ShapeWorkspace=data, DataWorkspace=norm)
data = DivideMD(LHSWorkspace=data, RHSWorkspace=norm_replicated)

scale = 1

data.setSignalArray(data.getSignalArray() / scale)
data.setErrorSquaredArray(data.getErrorSquaredArray() / scale**2)

data_norm = ConvertHFIRSCDtoMDE(data,
                                Wavelength=wavelength,
                                ObliquityParallaxCoefficient=cop,
                                MinValues='{},{},{}'.format(*min_values),
                                MaxValues='{},{},{}'.format(*max_values))

bragg = PredictPeaks(InputWorkspace=data_norm,
                     ReflectionCondition=reflection_condition,
                     CalculateGoniometerForCW=True,
                     MinAngle=-90,
                     MaxAngle=90,
                     Wavelength=wavelength,
                     OutputType='Peak')
```
- Open data_norm in `SliceViewer`
- Overlay the `bragg` peaks workspace
- Look for peaks 
  - (-1 0 -6) or (1 0 6)
  - (-1 0 -5) or (1 0 5)
  - (-1 0 -4) or (1 0 4)
  - (-1 0 -3) or (1 0 3)

> these are the peaks that were originally missing due to a limited q space search radius in DetectorSearcher.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/DetectorSearcher.cpp [hunk 1]
```diff
@@ -146,16 +151,22 @@ DetectorSearcher::DetectorSearchResult DetectorSearcher::searchUsingInstrumentRa
 DetectorSearcher::DetectorSearchResult DetectorSearcher::searchUsingNearestNeighbours(const V3D &q) {
   const auto detectorDir = convertQtoDirection(q);
   // find where this Q vector should intersect with "extended" space
-  const auto neighbours = m_detectorCacheSearch->findNearest(Eigen::Vector3d(q[0], q[1], q[2]), 5);
-  if (neighbours.empty())
+  // NOTE: increase the Nneighbors from 5 to 11 to cover a wide extended space
+  const auto neighbours = m_detectorCacheSearch->findNearest(Eigen::Vector3d(q[0], q[1], q[2]), 11);
```
Comment (wdzhou) [inline Framework/API/src/DetectorSearcher.cpp:155]: it is a magic number!

**Review by wdzhou — APPROVED** [review]
It looks good to me. :shipit:

**Comment by DavidFair** [comment]
I haven't reviewed this yet, but it looks like the system test `WISHSingleCrystalPeakPredictionTest` may have a legitimate failure that needs to be resolved

**Comment by RichardWaiteSTFC** [comment]
> I haven't reviewed this yet, but it looks like the system test `WISHSingleCrystalPeakPredictionTest` may have a legitimate failure that needs to be resolved

This is a flaky test that has failed on a few PRs. @DannyHindson has got a PR #31758 to increase the memory required which we hope will fix it. I'm not saying this PR definitely doesn't brake it though...I guess we'll know if it passes if you re-run it a couple of times. I can have a test if you like @DavidFair?

**Comment by DavidFair** [comment]
In this case though it looks like the number of spectra per peak has changed, which isn't typical of a flaky test?

**Comment by KedoKudo** [comment]
> In this case though it looks like the number of spectra per peak has changed, which isn't typical of a flaky test?



> In this case though it looks like the number of spectra per peak has changed, which isn't typical of a flaky test?

@DavidFair rebuilding seems to help resolve this issue.  Could you merge this PR? Thanks :pray:

**Review by DavidFair — APPROVED** [review]
LGTM, it was an unreliable test but it looked a lot like a genuine failure :confused:, sorry about that

**Comment by DannyHindson** [comment]
I think the ISIS_WISHSingleCrystalReduction.WISHSingleCrystalPeakPredictionTest was skipped on the most recent run of the Jenkins job. It's got a "memoryrequired" condition in the system test that means it often gets skipped. So it may be the case that it is broken and it just didn't get run on the last retry
I've made a small change to the memory required setting recently which is why I'm included on this PR

**Comment by wdzhou** [comment]
The compromise is to expose the number of nearest neighbors in `DetectorSearch` for customized setup.

**Comment by RichardWaiteSTFC** [comment]
I will update the reference file for the test in #31772

**Pull request merged**